### ISSUE TITLE
[Chrome] All the Externals by greasyfork.org of my installed scripts have become 0 byte (most probably because the weekly Externals update check 2 days ago was receiving empty pages for requests to Greasyfork scripts with a pinned version )

### ISSUE THREAD

**Issue by darkred** [Tampermonkey/tampermonkey#518]
_Using Tampermonkey 4.5 in Chrome 64, win10x64._ 

Today I noticed that a lot of my installed scripts in Chrome would no longer work. 
So I checked their settings | 'Externals' tab, and found that **for ALL my scripts that have `@require` rules via greasyfork that had a 'Last Updated' entry of "Fri Mar 02 2018" , their Externals had become 0 byte !!!** 
_All other Externals were still ok, e.g. those by GitHub:_
![2018-03-04_215338](https://user-images.githubusercontent.com/723651/36949880-8bba2364-1ff6-11e8-8317-e0557d7d4c95.jpg) 
![2018-03-04_215234](https://user-images.githubusercontent.com/723651/36949854-64dcb5e0-1ff6-11e8-9fce-0eac3d1e7a1c.jpg)

I have in TM settings | Externals | the 'Update' interval set as 'Weekly' (default value).

I suspect that this issue has happened after the weekly Externals update check was receiving empty pages for requests to Greasyfork scripts with a pinned version 
because some days ago I noticed that Externals URLs that point to specific versions, e.g. https://greasyfork.org/scripts/21927-arrive-js/code/arrivejs.js?version=139586 would return empty pages <sup>[2]</sup>.
And, so two days ago, 3/2/2018, when the weekly automatic updates check for Externals was executed,
it was retrieving empty pages, so I suppose that TM wrongly replaced them all with 0 byte content.

[1] <sup> (that's why I've [changed](https://greasyfork.org/en/scripts/34348-userstyles-org-css-highlighter/diff?v1=254630&v2=237602) the require rules for a script of mine 5 days ago to non-pinned versions)</sup>

**Comment by Mottie** [comment]
>[greasyfork] would return empty pages

This issue should now be fixed, see https://github.com/JasonBarnabe/greasyfork/issues/539.

**Comment by darkred** [comment]
@Mottie 
Yes, the issue no longer occurs. Thank you for the link.

---

@derjanb 
Now, the problem is that I cannot find a way to automatically make all these 0 byte Externals be replaced with the proper files:
I've already tried changing in Settings the Externals update interval from _Weekly_ to _Always_,
and closed + restarted Chrome: I noticed that the 'Last updated' timestamps in the Externals tab is changed with the current one,
but unfortunately they all remain 0 byte: [screnshot](https://user-images.githubusercontent.com/723651/37106618-f2c044c6-223a-11e8-9c19-e8937ccfdbc3.jpg)

I guess the only solution would be: 
- either re-install all scripts manually (out of the question)
- or export my scripts to ZIP, uninstall(in order to uninstall all scripts) + reinstall TM, and then re-import the ZIP. 

Or, is there another way, @derjanb ?

<br/>

And, in any case, I suggest that TM should prevent such situations, i.e. not replace the Externals with 0 byte, whenever the source site wrongly serves empty pages.

**Comment by derjanb** [comment]
> I've already tried changing in Settings the Externals update interval from Every Week to Always, and closed + restarted Chrome

In all cases the script needs to run in order to update the resources. If this happens then setting the interval to "Always" should work as well as manually deleting the resources and run the script again.

If the resource is still not updating for you while loading the URL manually shows a non-blank page, then please open [Tampermonkey's background page](http://tampermonkey.net/faq.php#Q600), select the network tab, try to refresh the resource again and post a screenshot of the network request details. Thanks.

> And, in any case, I suggest that TM should not replace the Externals with 0 byte, whenever the source site wrongly sends empty pages.

If the pages sends a HTTP response code that suggests that everything is fine, then there is no way to detect "wrong" blank pages.

**Comment by darkred** [comment]
Thank you for replying.

> In all cases the script needs to run in order to update the resources. If this happens then setting the interval to "Always" should work as well as manually deleting the resources and run the script again.
> 
> If the resource is still not updating for you while loading the URL manually shows a non-blank page, then please open Tampermonkey's background page, select the network tab, try to refresh the resource again and post a screenshot of the network request details. Thanks.

Unfortunately this didn't work.
I set it to _Always_ and loaded https://github.com/, a page where most of the installed scripts that have some 0 byte Externals run on: unfortunately the files remained 0 byte.
All the requests containing `?version=` in their URL return a status of 304 ( [screenshot_1](https://user-images.githubusercontent.com/723651/37118420-d649e8ca-225c-11e8-84f3-d65dbde5718b.jpg) )
Here is a screenshot of the 'Headers' tab for the 1st request: ( [screenshot_2](https://user-images.githubusercontent.com/723651/37118572-560ea398-225d-11e8-98ce-1cd8e050bf8b.jpg) )
Clicking the 'Preview' tab shows the full script ( [screenshot_3](https://user-images.githubusercontent.com/723651/37118514-2b9d2c24-225d-11e8-9f37-002579ca2838.jpg) )

**Comment by derjanb** [comment]
[304](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304) means "Not modified". That's why the browser uses the cached (zero size) copy and forwards this to Tampermonkey. Please clear your browser cache and everything should work again.

**Comment by darkred** [comment]
> Please clear your browser cache and everything should work again. 

Unfortunately it doesn't work:

I opened Settings|Clear browsing data|Basic|Time range: All time| ticked only the "Cached images and files" and pressed CLEAR DATA:
unfortunately: all requests returned 304.

Then I ticked the checkbox "Disable cache "in Tampermonkey's background page: now all requests returned 200, (and the "Preview" tabs show full scripts)
but still the files remain 0 bytes.

**Comment by darkred** [comment]
How about adding an extra Externals update interval setting named "Forced", that would force-download/update all Externals, regardless of the response status codes by the server? 
It would be set one time, i.e. the user would set the Externals update interval to 'Forced' once, and it would apply only on the current browser session -- in the next browser session TM would be re-set to its previous value.

**Comment by darkred** [comment]
@derjanb Is there any hope/change for providing a way to fix this this in a coming TM beta?
Or should I close this issue and just uninstall TM (in order all scripts to be removed) + try to reinstall TM re-importing the scripts from ZIP backup?

**Comment by derjanb** [comment]
@darkred Are there any messages at the background console messages when an external is updated. I don't see a reason why TM shouldn't use the retrieved request data. Have you maybe changed the "Subresource Integrity" option to "Enforce"?

**Comment by darkred** [comment]
@derjanb 
> Are there any messages at the background console messages when an external is updated. 

For reference, as I've said, I have set the Externals update interval set to _Always_, so, whenever I load https://github.com/ the 0 byte Externals should be updated to the expected files.
But, no, I get no error at all when loading https://github.com/ ([screenshot](https://user-images.githubusercontent.com/723651/37566305-e7b50d72-2abf-11e8-9ac1-e50f1da5d77e.jpg)) 

I have already mentioned that I've tried:
- opening Chrome's Settings|Clear browsing data|Basic|Time range: All time| ticked only the "Cached images and files" checkbox and pressed CLEAR DATA, then reloaded https://github.com/ : 
unfortunately, all requests return 304.

- ticking the checkbox "Disable cache "in Tampermonkey's background page|'Network' tab, then reloaded https://github.com/ :
Now all requests return 200 ([screenshot](https://user-images.githubusercontent.com/723651/37126526-121245c4-227a-11e8-9228-53083fbdf172.jpg)) (and the "Preview" tab for all requests shows full scripts) but still the Externals remain 0 bytes.

**Is there anything more to try  regarding the clearing of the browser cache ?**

&nbsp;&nbsp;&nbsp;


> Have you maybe changed the "Subresource Integrity" option to "Enforce"?

No, the option remains with its default value: `Validate if possible`. ([screenshot](https://user-images.githubusercontent.com/723651/37566006-da484d74-2abb-11e8-81d4-51870bb13869.jpg))

**Comment by derjanb** [comment]
Can you please set the log level to "debug" and try again? You then should see some output similar to this:

![image](https://user-images.githubusercontent.com/767504/37567106-53f2905c-2ac2-11e8-9ce3-2d972c3a3c93.png)

**Comment by darkred** [comment]
Yes. of course. Here: [output log](https://github.com/Tampermonkey/tampermonkey/files/1822766/dhdgffkkebhmkfjojejmpbldmpobfkfo-1521384142370.log)

**Comment by derjanb** [comment]
And when you click at the arrow to open "resource" how does the item look? Ist there a "content" property and what is its value?

**Comment by darkred** [comment]
Yes, there's a "content" property for all entries, and its value in all cases is the wanted script (see e.g. the tooltip in the bottom of the screenshot).
Here it's a screenshot: 
![2018-03-18_172324](https://user-images.githubusercontent.com/723651/37567546-39a4d160-2ad1-11e8-8e40-8f24401c3c53.jpg)

**Comment by darkred** [comment]
@derjanb 
One last thought, please:
I've noticed in Console for any of my installed scripts that have one 0 byte External, 
that whenever I click the Edit button in its Externals tab I get:
```
options.html#nav=b5c1c9ac-8f90-4235-987e-5c20c1a0e7ce+externals-b5c1c9ac-8f90-4235-987e-5c20c1a0e7ce-https://greasyfork.org/scripts/28721-mutations/code/mutations.js?version=234970:1 Error in event handler for (unknown): TypeError: Cannot read property 'first' of null
    at new $f (chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/vendor/cm/lib/codemirror.js:11:71)
    at new E (chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/vendor/cm/lib/codemirror.js:173:357)
    at new window.MirrorFrame (chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/vendor/cm/mirrorframe.js:3:415)
    at Object.<anonymous> (chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/layout/default/options.js:108:402)
    at Array.f.check (chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/promise.js:1:322)
    at Object.try_resolve (chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/promise.js:3:383)
    at Object.resolve (chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/promise.js:4:79)
    at chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/layout/default/options.js:108:69
    at chrome-extension://dhdgffkkebhmkfjojejmpbldmpobfkfo/layout/default/options.js:162:316
```
Of course, this only happens for 0 byte Externals, not normal ones.

So, maybe these 0 byte files have actually become non-existent (null), i.e. only their file name remains in TM database,
and so they can't be updated to full filesize by TM. unless I manually delete them(those 0 byte Externals), one by one ??

**Comment by derjanb** [comment]
Finally I found the issue. 😅 Thank for your help on debugging this! The issue is fixed at TM Beta 4.6.5741, but not at 4.5 (yet).

You should be able to fix the issue by manually deleting every single zero byte external or by exporting all scripts (please check if the import works at this point), and then delete all scripts and import them again.

**Comment by darkred** [comment]
I'm glad!

Yes, I managed to fix the issue by I exporting all scripts(incl. TM settings) from TM 4.5 , deleted all scripts and imported them again to TM beta 4.5.
The import worked fine, and all externals I checked are ok.

PS. I decided to permanently switch to TM beta version (so I installed TM beta 4.6.5741 and imported the scripts), like I do with Firefox,
